### PR TITLE
Support BaguaStrategy with external implementation

### PR DESCRIPTION
**Pull request by woqidaideshi**
## What does this PR do?

The Bagua Strategy moves to an external repo where it will be maintained: https://github.com/Lightning-AI/lightning-Bagua. [A corresponding package: lightning-bagua](https://pypi.org/project/lightning-bagua/) have been published to PyPi from which users can install the integration.

-  Replaces the implementation with the one in the lightning-bagua package.
-  Registeres the strategy="bagua" as a strategy if the package is installed
-  Produces an error message if you request the strategy but don't have the package installed

<details>
  <summary><b>Before submitting</b></summary>

- [ ] Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

**Comment by woqidaideshi** [comment]
@Borda @awaelchli  Could you help us to review this PR? Thanks!

**Review by Borda — APPROVED** [review]
let's also add one-age documentation in here:
https://github.com/Lightning-AI/lightning/tree/master/docs/source-pytorch/advanced/third_party

**Comment by woqidaideshi** [comment]
> let's also add one-page documentation here: https://github.com/Lightning-AI/lightning/tree/master/docs/source-pytorch/advanced/third_party

@Borda Thanks for your suggestion. I have added one-page documentation for the introduction of BaguaStrategy.

File: docs/source-pytorch/advanced/third_party/bagua.rst [hunk 1]
```diff
@@ -0,0 +1,49 @@
+:orphan:
+
+###########
+Bagua
+###########
```
Comment (Borda) [inline docs/source-pytorch/advanced/third_party/bagua.rst:5]:
```suggestion
#####
Bagua
#####
```

File: docs/source-pytorch/advanced/third_party/bagua.rst [hunk 2]
```diff
@@ -0,0 +1,49 @@
+:orphan:
+
```
Comment (Borda) [inline docs/source-pytorch/advanced/third_party/bagua.rst:2]:
```suggestion
```

**Review by awaelchli — APPROVED** [review]

File: src/lightning/pytorch/utilities/imports.py [hunk 3]
```diff
@@ -30,6 +30,7 @@
 _RICH_AVAILABLE = package_available("rich") and compare_version("rich", operator.ge, "10.2.2")
 _TORCHVISION_AVAILABLE = RequirementCache("torchvision")
 _LIGHTNING_COLOSSALAI_AVAILABLE = RequirementCache("lightning-colossalai")
+_LIGHTNING_BAGUA_AVAILABLE = RequirementCache("lightning-bagua")
```
Comment (carmocca) [inline src/lightning/pytorch/utilities/imports.py:33]: These 2 should be defined in the accelerator connector, which is the only file that uses it.

**Review by carmocca — APPROVED** [review]

File: src/lightning/pytorch/utilities/imports.py (hunk 3 above)
Comment (woqidaideshi) [reply src/lightning/pytorch/utilities/imports.py:33]: @carmocca In fact, there are other files that use `_LIGHTNING_COLOSSALAI_AVAILABLE`, such as `src/lightning/pytorch/callbacks/gradient_accumulation_scheduler.py` and `tests/tests_pytorch/callbacks/test_gradient_accumulation_scheduler.py`. However, `_LIGHTNING_BAGUA_AVAILABLE` is currently only used in `src/lightning/pytorch/trainer/connectors/accelerator_connector.py`.

@Borda @carmocca  Any suggestions to fix it?

Comment (carmocca) [reply src/lightning/pytorch/utilities/imports.py:33]: It's okay. I'll look into this in a follow-up

**Comment by woqidaideshi** [comment]
@lantiga   @williamFalcon   Could you help us to review this PR? Thanks!

**Review by lantiga — APPROVED** [review]
Looks good to me, there's still an html workflow failing @Borda

**Pull request merged**